data/calculated_parser/functions.py

```diff
@@ -69,11 +70,293 @@ def sspeed(depth, latitude, temperature, salinity):
+
+    if lower_subset.size == 0:
+        return np.nan
+
+    # Find min and max value of subset
+    min_value = np.nanmin(lower_subset)
+    max_value = np.nanmax(lower_subset)
+
+    # Determine if Critical Depth exists
+    if max_value < sld_value:
+        return np.nan
+
+    # Find closest idx to sld_value
+    idx = np.abs(lower_subset - sld_value).argmin()
+
+    # Find size of lower subset
+    subset_size = lower_subset.size - np.count_nonzero(np.isnan(lower_subset)) - 1
+
+    # Shift to get index for non subset
+    cd_idx = total_idx - (subset_size - idx)
+
+    return cd_idx
+
+def cd_interpolation(cd_idx, sld_idx, speed_point, depth):
+
+    # Now that we have the nearest critical depth idx we must perform linear interpolation
+    def linearInterp(x1, y1, x2, y2, x):
+        """
+        Finds the linear interpolation given 2 points
+        """    
+        y = y1 + ((x - x1) * ((y2 - y1) / (x2 - x1)))
+        return y
+
+    sld_value = speed_point[sld_idx]
+    cd_value = speed_point[cd_idx]
+    cd_depth = depth.values[cd_idx]
+
+    if cd_value < sld_value:
+        
+        # Find 2 Points
+        cd_value_1 = cd_value
+        cd_value_2 = speed_point[cd_idx + 1]
+
+        cd_depth_1 = depth.values[cd_idx]
+        cd_depth_2 = depth.values[cd_idx + 1]
+
+        cd_depth = linearInterp(cd_value_1, cd_depth_1, cd_value_2, cd_depth_2, sld_value)
+
+    elif cd_value > sld_value:
+        
+        # Find 2 Points
+        cd_value_1 = speed_point[cd_idx - 1]
+        cd_value_2 = cd_value
+
+        cd_depth_1 = depth.values[cd_idx - 1]
+        cd_depth_2 = depth.values[cd_idx]
+
+        cd_depth = linearInterp(cd_value_1, cd_depth_1, cd_value_2, cd_depth_2, sld_value)
+
+    return cd_depth
+
+def criticaldepth(depth, lat, lon, temperature, salinity):
+    """
+    Finds the next location of the sonic layer depth.
+
+    Parameters:
+    sspeed: Speed of Sound
+    latitude: The latitude(s) in degrees North
+    """
+
+    speed = sspeed(depth, lat, temperature, salinity)
+    speed = speed.transpose()
+    sld = 0
+    sca = 0
+    for x in range(speed.shape[0]):
+        for y in range(speed.shape[1]):
+            if (speed[x,y].size - np.count_nonzero(np.isnan(speed[x,y]))) != 0:
+                speed_point = speed[x,y]
+                sca_idx = find_sca_idx(speed_point)
+                if not np.isnan(sca_idx):
+                
+                    sld_idx = find_sld_idx(sca_idx, speed_point)
+
+                    if not np.isnan(sld_idx):
+                        cd_idx = find_cd_idx(sca_idx, sld_idx, speed_point)
+                        if not np.isnan(cd_idx):
+                            #cd_value = speed_point[cd_idx]
+                            #cd_depth = depth.values[cd_idx]
+
+                            # Now that we have the nearest critical depth idx we must perform linear interpolation
+                            cd_depth = cd_interpolation(cd_idx, sld_idx, speed_point, depth)
```

Comment:
Here we interpolate between the sonic layer depth and the depth of the grid point closest to the critical depth
Shouldn't we do something like this:

``` python
# If a critical depth exists, speed should be increasing with depth ( right??)
# So if speed_point[cd_indx] < sld_value -> We need to look deeper to cross critical depth - use cd_idx+1
# if speed_point[cd_idx] > sld_value - We need to look shallower to cross critical_depth - use cd_idx - 1
if  speed_point[cd_idx] < speed_point[sld_idx]
    cd_depth = cd_interpolation(cd_idx, cd_idx+1, speed_point, depth) # Question - can we hit a nan here??
else
    cd_depth = cd_interpolation(cd_idx,  cd_idx-1, speed_point, depth)
```

I think it would be more accurate to interpolate around the neighbourhood of the depth closest to the critical layer depth, rather than look all the way back up to the sonic layer depth. But I need to check my assumption that the sound speed is increasing at this point...

We can chat more next week.

Reply:
I think this is what I'm doing, I'm not using the critical depth, just the sound speed at that value.
I'm looking for the depth at which the sonic layer depths sspeed occurs by looking at the depth levels before and after the expected point, and performing a linear interpolation with those 2 depth levels.

Let me know if I misunderstood.

Reply:
Ok, nevermind... I took a closer look at the interpolation function and you do what I suggested. My bad!!